### ISSUE TITLE
Question: Verification always fail 

### ISSUE THREAD

**Issue by Lisalora** [maticnetwork/miden#342]
Hello everyone, I'm new here, trying to use the `next` branch of midenVM . 
I have a question which is probably a bit silly. I tried to use Miden (`next` branch) and found that no matter what program I use, the verification result is `Err (VerifierError (InconsistentOodConstraintEvaluations)).`

If the program has `push` operation, i might have the assertion error:
```bash
thread 'main' panicked at 'assertion failed: `(left == right)`
left: `QuadExtension(BaseElement(9801321997299852158), BaseElement(10889698089612978941))`,
right: `QuadExtension(BaseElement(1), BaseElement(0))`', /Users/lisalora/.cargo/git/checkouts/miden-4241ff496f7b0a1f/8b07d87/processor/src/trace/utils.rs:188:9
```
The program i used is as follows:

```rust
use miden;
fn main() {
    // or `begin add end`
    let program = "
    begin 
        push.1 
    end";
  
    let options = &miden::ProofOptions::with_96_bit_security();
    let assembler = miden::Assembler::new(true);
    let program = assembler.compile(&program).unwrap();
    let inputs = miden::ProgramInputs::new(&[1,2,3],&[],vec![]).unwrap();

    let num_stack_outputs = 5;
    let proof = miden::prove(&program, &inputs, num_stack_outputs, options);
    let (output, proof) = proof.unwrap();
    let verify_result = miden::verify(program.hash(), &[], &output, proof);
    println!("verify_result is {:?},output is {:?}",verify_result,output);
}
```

Could somebody possibly tell me whether it is my fault or i used the wrong commit of miden?
If the latest commit cannot be verified, what is the latest workable commit? (I want to try some new function of `next`(such as read-write random-access memory, etc.), rather than the `main` version) 

Thanks for any reply

**Comment by bobbinth** [comment]
Hi there! Not sure if this will solve this, but one issue that I'm seeing is that you are not using the same inputs for generating and verifying the proof. Try changing:
```Rust
let verify_result = miden::verify(program.hash(), &[], &output, proof);
```
To:
```Rust
let verify_result = miden::verify(program.hash(), &[1, 2, 3], &output, proof);
```
And see if that fixed it.

**Comment by Lisalora** [comment]
> Hi there! Not sure if this will solve this, but one issue that I'm seeing is that you are not using the same inputs for generating and verifying the proof. Try changing:
> 
> ```rust
> let verify_result = miden::verify(program.hash(), &[], &output, proof);
> ```
> 
> To:
> 
> ```rust
> let verify_result = miden::verify(program.hash(), &[1, 2, 3], &output, proof);
> ```
> 
> And see if that fixed it.

sorry... my bad. But after fill in the `input[1,2,3]`, the verification result is still `Err(VerifierError(InconsistentOodConstraintEvaluations))`.

**Comment by bobbinth** [comment]
Seems like there are a couple of other things happening here:

1. Miden VM expects that stack depth at the end program execution is exactly 16. We should document this much better and also show a much better error message if the stack depth is not 16 by the end of the program.
2. It seems like setting number of outputs to greater than 1 also causes issue. This is probably a bug that we'll need to investigate.

If you set your program to something like:
```
begin
    push.1
    movup.5
    drop
end
```
And set `num_stack_outputs` to `1` - it should work.

**Comment by bobbinth** [comment]
Looking a bit more into the number of outputs bug - the issue is that the verifier interprets the output vector in a different order (so, if we reverse the output vector before feeding it to the verifier, everything should work). We should make sure to interpret the order of inputs and outputs consistently in the prover and the verifier and also create a test which covers the cases with more than one output.

**Comment by Lisalora** [comment]
> Looking a bit more into the number of outputs bug - the issue is that the verifier interprets the output vector in a different order (so, if we reverse the output vector before feeding it to the verifier, everything should work). We should make sure to interpret the order of inputs and outputs consistently in the prover and the verifier and also create a test which covers the cases with more than one output.

Great! After reversing the output, it works!

**Comment by Lisalora** [comment]
> Seems like there are a couple of other things happening here:
> 
> 1. Miden VM expects that stack depth at the end program execution is exactly 16. We should document this much better and also show a much better error message if the stack depth is not 16 by the end of the program.
> ........


Another question about the `1`, how could the stack depth becomes exactly 16 at the end of the program?(after adding the two operations: `movup.5 & drop`)


I did another test for `dup.0`, the program shows next will fail (the origin_input on stack is [22,22])
```
begin
     dup.0
end
```
But the program adding `movup.5 & drop` works fine, I would appreciate it if you tell me what happened here.

**Comment by grjte** [comment]
@Lisalora I agree with @bobbinth that we should make some adjustments for usability, so thanks for bringing this up! For now, here's an explanation of what's going on.

TL;DR For any stack depth increase that occurs during program execution, there must be a corresponding stack depth decrease before the execution ends. (e.g. `dup.0` must be accompanied by a `drop` at the end)

In more detail -

MIden VM allows a practically unlimited stack depth, but it does this by using an "overflow table" for all items deeper than the 16th spot. For this construction we have two requirements:

- The stack must have a minimum depth of 16 during execution, which we do to simplify constraints. This means that if fewer than 16 inputs are provided, the rest of it is padded with zeros before execution begins. That means the initial stack when your program was started was actually [22,22,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
- The "overflow table" be empty at the end of the program (functionally, this means that the stack will have exactly 16 elements at the end.)


So what's happening here is that the stack has a depth of 16 when execution starts, then you call `dup.0` which increases the depth to 17 and pushes the deepest item onto the overflow table. In order for execution to succeed, you just need to call `drop` again, which will remove the top item and move the other item out of the overflow table and back onto the stack.

You can read more about the design [here](https://maticnetwork.github.io/miden/design/stack/main.html)

**Comment by grjte** [comment]
@Lisalora We have a stdlib function that can be called at the end of execution to resolve the stack depth issue: https://maticnetwork.github.io/miden/user_docs/stdlib/sys.html

**Comment by Lisalora** [comment]
> @Lisalora We have a stdlib function that can be called at the end of execution to resolve the stack depth issue: https://maticnetwork.github.io/miden/user_docs/stdlib/sys.html

Thanks!! An unexpected error comes when trying to figure out how to use `finalize_stack`:
The program with `gte not` works great:
```
   begin
        push.1
        push.2
        gte not
        exec.sys::finalize_stack
    end
```
However, replace `gte not` with `lt` (which sounds equal) leads to an unexpected error. 
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `QuadExtension(BaseElement(8775471544908860125), BaseElement(15028436441736881570))`,
 right: `QuadExtension(BaseElement(1), BaseElement(0))`'
```

This confuses me a little, should I also lookup for the reason in the design [here](https://maticnetwork.github.io/miden/design/stack/main.html)

**Comment by bobbinth** [comment]
This could be a bug inside `finalize_stack` procedure. We'll need to investigate. Thank you for finding all these!

**Comment by 0xKanekiKen** [comment]
> This fails with a debug assertion failure while building the Range checker's p1 auxiliary column (it's not ONE where it's expected to be):

I think it might be because of the fact that we are doing an `u32overflowing_sub` on 1 & 2 here (1 - 2). The `lo` & `hi` are result(1-2) in 2's complement as u32 and 0. Decomposing the result as u32 in two 16-bit numbers will yield two U16MAX - 1.
When we are building the running product column for 16-bit section of the range checker table, we are stopping the iteration right before U16MAX - 1 due to which component of U16MAX - 1 are not getting incorporated into the running product and that's why p1 is not coming out to be 1. 

We will see this error for the following case as well where we are getting two U32MAX - 1 on decomposing the result in two 16-bits number:
```
   begin
        push.1
        push.U32MAX - 2
        u32overflowing_add
        drop
        drop
    end
```

In the case of `gte`, the reason for not seeing this issue is because here we are doing 2-1 which will not decompose into U16MAX - 1 during range check operations. Although now I'm seeing a new error when I tried to incorporate U16MAX-1 into the `p1` product column.
```
thread 'examples::testing::test_get_example' panicked at 'trace does not satisfy assertion aux_trace(5, 1022) == (1, 0)', /home/miden/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.0/src/trace/mod.rs:151:17
```

**Comment by grjte** [comment]
@0xKanekiKen thanks for your insights on this! I'll dig into it more today

**Comment by bobbinth** [comment]
Closing this as after #346 and #353 the bugs should be fixed. Figuring out the best way to simplifying stack management is moved to #347.

@Lisalora - thank you again for bringing these up! Feel free to open other issues if you come across anything else.
